### PR TITLE
Exceptions

### LINKED ISSUE #55: ExceptionTest::test_defunct_processes_kill sometimes fails on Windows machines
tests/test_pool.py::ExceptionTest::test_defunct_processes_kill sometimes deadlocks when running the tests with GitHub actions. This only started to occur quite recently. The Python version and test arguments don't seem to have an influence.

Failed runs:
- https://github.com/Slimmer-AI/mpire/runs/7496816178?check_suite_focus=true
- https://github.com/Slimmer-AI/mpire/runs/7302594707?check_suite_focus=true
- https://github.com/Slimmer-AI/mpire/runs/7298408511?check_suite_focus=true

Will need to debug this on Windows.

### LINKED ISSUE #56: Deadlock when using dill and an exception is thrown
When using dill and an exception is thrown in one of the worker processes the main process deadlocks.

reproducable example:
```
from mpire import WorkerPool

# Create some fake data. It should raise when x=10, y=52, z=0
data = [(x, y, z) for x, y, z in zip(range(0, 100), range(42, 142), range(10, -90, -1))]
with WorkerPool(n_jobs=5, use_dill=True) as pool:
    for res in pool.imap(lambda x, y, z: x*y/z, data):
        print(res)

```

Wwhen setting use_dill = False it works like a charm. Unfortunaeley in my case I have to use dill to pickle lambda functions.

### LINKED ISSUE #57: Progress bar description
MPIRE uses TQDM as a progress-bar, which is very nice. However, it would be great if there was the possibility to somehow control the progress-bar with the parameters that are available, e.g., the `desc`ription. We have several progress bars in our app and it would be nice to see not only that something is happening, but also _what_ is happening.

Do you think that would be a possible expansion of MPIRE? Is there another way how you can configure TQDM? If yes, where can I find this information in the documentation?

### LINKED ISSUE #58: Exceptions with mandatory arguments are not reraised correctly
If some exception with positional arguments is being raised by a subprocess, this exception is not being raised correctly. Most of the times it even causes deadlocks in my attempts to create minimum reproducable examples.

In case the exception takes exactly one argument, the exception message might be wrong. In case it takes more than one mandatory positional arguments the reraising will fail, because in following line the exception is assumed to take exactly one argument (the exception message)
https://github.com/Slimmer-AI/mpire/blob/0041051aaf4acf83e4c7e106a4f651d7ee6f3f26/mpire/pool.py#L752

Minimal examples:

```
from mpire import WorkerPool

class MyException1(Exception):
    def __init__(self, a, b):
        self.a = a
        self.b = b
        super().__init__(f"a={self.a}, b={self.b}")

class MyException2(Exception):
    def __init__(self, a):
        self.a = a
        super().__init__(f"a={self.a}")

def work1(x):
    raise MyException2(x)

def work2(x):
    raise MyException2(x, x)

with WorkerPool(n_jobs=5, use_dill=False) as pool:
    for res in pool.imap(work1, range(10)):
        if isinstance(res, Exception):
            raise res
        else:
            print(res)
```

### PR DESCRIPTION
**Pull request by sybrenjansen**
Fixes #55 
Fixes #56 
Fixes #57 
Fixes #58

### CODE REVIEW

**Comment by sybrenjansen** [comment]
Ready for review @basroelenga

File: mpire/params.py [hunk 1]
```diff
@@ -221,11 +230,49 @@ def check_map_parameters(pool_params: WorkerPoolParams, iterable_of_args: Union[
...
+
+    # Set some defaults and overwrite others
+    progress_bar_options["total"] = n_tasks
+    progress_bar_options["leave"] = True
+    progress_bar_options.setdefault("position", 0)
+    progress_bar_options.setdefault("dynamic_ncols", True)
+    progress_bar_options.setdefault("mininterval", 0.1)
+    progress_bar_options.setdefault("maxinterval", 0.5)
+
+    # Check that all progress bar options are properly formatted. We disable the __del__ method of the tqdm class to
```
Comment (sybrenjansen) [inline mpire/params.py:267]: Shit, the `__del__` comment is an old comment that should be removed

File: mpire/comms.py [hunk 2]
```diff
@@ -132,6 +132,7 @@ def init_comms(self) -> None:
 
         # Exception related
         self._exception_queue = self.ctx.JoinableQueue()
+        self._exception_received = self.ctx.Event(), self.ctx.Event()
```
Comment (michielvandesteeg) [inline mpire/comms.py:135]: Could be slightly clearer by either adding a comment specifying the different events or simply adding one more variable for the progress bar event.

Still easy to understand with a quick scroll or ctrl+f so fine as is, just at first glance it's not clear.

File: mpire/comms.py [hunk 3]
```diff
@@ -410,11 +411,26 @@ def get_exception(self) -> Tuple[type, str]:
...
 
-    def task_done_exception(self) -> None:
+    def task_done_exception(self, progress_bar: bool = False) -> None:
         """
         Signal that we've completed a task for the exception queue
+
+        :param progress_bar: Whether the exception was handled for the progress bar
         """
         self._exception_queue.task_done()
+        self._exception_received[0].set() if not progress_bar else self._exception_received[1].set()
```
Comment (michielvandesteeg) [inline mpire/comms.py:423]: If you keep _exception_received as a tuple:

```
self._exception_received[postgres_bar].set()
```

For fanciness points! :P

**Review by michielvandesteeg — APPROVED** [review]
Looks good to me. See whether you want to do anytthing about the event tuple comment.

File: docs/changelog.rst [hunk 4]
```diff
@@ -4,8 +4,9 @@ Changelog
 Dev
 ---
 
-* Fixed a bug where a deadlock could occur when an exception was raised from a worker. When the exception payload was
-  bigger than the buffer, the worker could exit before everything was send over the queue (`#56`_)
+* Added Python 3.10 support
+* Fixed a bug where a worker could exit before an exception was entirely send over the queue, causing a deadlock
```
Comment (michielvandesteeg) [inline docs/changelog.rst:8]: Old and nitpicky, but "sen**t**"

File: mpire/comms.py (hunk 3 above)
Comment (sybrenjansen) [reply mpire/comms.py:423]: Although it work in this case, I'm generally not a big fan of using a boolean as an integer. But I might change the tuple to two separate variables.

**Review by michielvandesteeg — COMMENTED** [review]
Changes in new commits look good as well though builds are failing, is that related to the process lock?

**Comment by sybrenjansen** [comment]
Just ignore the commits for now, I'm trying to debug windows issues. I can't reproduce them on my on Windows machine, so abusing github a bit for this

**Comment by sybrenjansen** [comment]
@michielvandesteeg It's ready now. Just need your final approval ;)

**Review by michielvandesteeg — APPROVED** [review]

**Comment by michielvandesteeg** [comment]
> Just ignore the commits for now, I'm trying to debug windows issues. I can't reproduce them on my on Windows machine, so abusing github a bit for this

Are all the debug messages intended to stay in or did you just add them because of this?

**Comment by sybrenjansen** [comment]
Debug messages are indeed to stay. They don't hurt anyone

**Pull request merged**